MDEV-12459 : Get temporary tables visible to the IS.tables for current connection

an3l opened the pull request:
This PR is related to the MDEV-12459 and all MDEVs that are caused by it (see JIRA).
Most of commits are reviewed, but further review is needed, like MDEV-28334 and MDEV-28453.
Closes PR#695.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=2189) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=2189) before we can accept your contribution.<br/>**2** out of **4** committers have signed the CLA.<br/><br/>:white_check_mark: bnestere<br/>:white_check_mark: an3l<br/>:x: mariadb-YuchenPei<br/>:x: montywi<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=2189) it.</sub>

an3l committed:
MDEV-28343: sys.create_synonym_db fails with ER_VIEW_SELECT_TMPTABLE when schema contains temporary tables

- MDEV-28342 raised the error in case temporary table shadows base table
- Now we are allowed to shadow base tables with temporary tables and
`sys.create_synonym_db()` can easily check for existance of temporary table and
ignore view creation, since it is not supported to create view from
temporary table.

Reviewed-by: <monty@mariadb.org>,
             <vicentiu@mariadb.org>

an3l commented:
@cvicentiu updated to 10.11 and rebased the commits, please review.

Commit:
MDEV-30188: Ensure all binlog* variables are visible as system variables
Turn the remaining three `binlog*` options binlog_do_db, binlog_ignore_db,
binlog_rows_event_max_size into global variables so that they can be
visible from the SQL user level. This is for audit / secure
configuration check purposes.

Create new MTR tests to make sure that the newly created global
variables can be visible from the command line interface.

Behavior before the code change:

    MariaDB [(none)]> SHOW GLOBAL VARIABLES WHERE
        -> Variable_name LIKE 'binlog_do_db' OR
        -> Variable_name LIKE 'binlog_ignore_db' OR
        -> Variable_name LIKE 'binlog_row_event_max_size';
    Empty set (0.001 sec)

Behavior after the code change:

    MariaDB [(none)]> SHOW GLOBAL VARIABLES WHERE
        -> Variable_name LIKE 'binlog_do_db' OR
        -> Variable_name LIKE 'binlog_ignore_db' OR
        -> Variable_name LIKE 'binlog_row_event_max_size';
    +---------------------------+-------+
    | Variable_name             | Value |
    +---------------------------+-------+
    | binlog_do_db              |       |
    | binlog_ignore_db          |       |
    | binlog_row_event_max_size | 8192  |
    +---------------------------+-------+
    3 rows in set (0.001 sec)

Note:

For `binlog_do_db` and `binlog_ignore_db`, we add a new class
`Sys_var_binlog_filter` to handle the dynamically-composable command line
options for `binlog_do_db` and `binlog_ignore_db`. Below
is the motivation:

When the users start the server with the option
    --binlog-do-db="database1" --binlog-do-db="database2"
The expected behavior is that the system should allow replication for
both `database1` and `database2`, which is the logic of the original
code.

However, when turning the variables into system variables, the
functionality does not exist any more, since system variables will only
handle the last occurrence of the option, and in this case, the system
will only be able to handle `database2`.

Copyright:

All new code of the whole pull request, including one or several files
that are either new files or modified ones, are contributed under the BSD-new
license. I am contributing on behalf of my employer Amazon Web Services, Inc.

LinuxJedi commented:
@an3l this PR includes too many things in one go and is difficult to review. Can you please break it up to something like one pull request per MDEV?

Also, correct me if I'm wrong, but I believe Monty developed the last patch in the series, but you appear to be marked as the author. If I am correct please fix the author for that commit.

an3l committed:
MDEV-12459 Get temporary tables visible to the IS.tables for current connection

Additionally fixes the bugs uncovered in:
  - `MDEV-28332: Alter on temporary table causes ER_TABLE_EXISTS_ERROR note`
    Since there is no `warning` issued by shadowing of base table, this MDEV
    is irrelevant. Keeping for review purposes and for future development
    in case shadowing is going to be implemented
  - `MDEV-28334: SHOW TABLE STATUS shows all temporary tables ignoring database and conditions`
  - `MDEV-28453: SHOW commands are inconsistent for temporary tables`

Reviewed by: <monty@mariadb.org>,
             <vicentiu@mariadb.org>

an3l committed:
MDEV-12459 Patch sysschema

This commit updates sysschema to work with the new behaviour of show
tables and information_schema.tables table showing temporary tables for
current connection.

Co-authored-by: Monty <monty@mariadb.org>
Reviewer: <vicentiu@mariadb.org>

an3l committed:
Cosmetic fixes

Reviewer: <vicentiu@mariadb.org>

mariadb-YuchenPei committed:
MDEV-26137 Improve import tablespace workflow.

Allow ALTER TABLE ... IMPORT TABLESPACE without creating the table
followed by discarding the tablespace.

That is, assuming we want to import table t1 to t2, instead of

CREATE TABLE t2 LIKE t1;
ALTER TABLE t2 DISCARD TABLESPACE;
FLUSH TABLES t1 FOR EXPORT;
--copy_file $MYSQLD_DATADIR/test/t1.cfg $MYSQLD_DATADIR/test/t2.cfg
--copy_file $MYSQLD_DATADIR/test/t1.ibd $MYSQLD_DATADIR/test/t2.ibd
UNLOCK TABLES;
ALTER TABLE t2 IMPORT TABLESPACE;

We can simply do

FLUSH TABLES t1 FOR EXPORT;
--copy_file $MYSQLD_DATADIR/test/t1.cfg $MYSQLD_DATADIR/test/t2.cfg
--copy_file $MYSQLD_DATADIR/test/t1.frm $MYSQLD_DATADIR/test/t2.frm
--copy_file $MYSQLD_DATADIR/test/t1.ibd $MYSQLD_DATADIR/test/t2.ibd
UNLOCK TABLES;
ALTER TABLE t2 IMPORT TABLESPACE;

We achieve this by creating a "stub" table in the second scenario
while opening the table, where t2 does not exist but needs to import
from t1. The "stub" table is similar to a table that is created but
then instructed to discard its tablespace.

We include tests with various row formats, encryption, with indexes
and auto-increment.

Signed-off-by: Yuchen Pei <yuchen.pei@mariadb.com>

montywi committed:
MDEV-28351 Assertion `this->file->children_attached' failed in ha_myisammrg::info

Reviewed-by: <vicentiu@mariadb.org>

cvicentiu commented on sql/sql_show.cc:
```diff
@@ -4800,7 +4805,8 @@ fill_schema_table_by_open(THD *thd, MEM_ROOT *mem_root,
                                    LEX_CSTRING *db_name,
-                                   LEX_CSTRING *table_name)
+                                   LEX_CSTRING *table_name,
+                                   bool is_temp= false)
```
Hi Anel!

I have identified a potential problem with this code.

Can you explain to me what is the behaviour of this function *before* your changes in the cases:

** ASSUME is_sequence is always false **

1a. `ha_table_exists` returns `true`;
2a. `ha_table_exists` returns `false`;

Then, can you explain to me what is the behaviour of this function when:

1b. `ha_table_exists` returns `true`, `is_temp == false`
2b. `ha_table_exists` returns `false`; `is_temp == false`
3b. `ha_table_exists` returns `true`, `is_temp == true`
4b. `ha_table_exists` returns `false`; `is_temp == true`

More specifically I'm interested in knowing what value is stored in `table->field[3]`.

an3l replied:
Hi @cvicentiu 
`fill_schema_table_names` fills table `IS.table` with data (in your specific question) from base table or temporary table. Before the patch temporary tables were not supported. Wit my patch I used the same function to show temporary tables. Adding `is_temp` allows to handle temporary table that are shadowing the base table (with the same name) and it is up to the the caller to set appropriate flags. We can set `dbug_assert` for situations that may be invoked by the caller (like last combination in below table).
Here is the output you asked:
  | ha_table_exists | is_temp | table->field[3]
-- | -- | -- | --
Before change | 1 | x | Base table
Before change | 0 | x | error
After change | 0 | 0 | error
After change | 0 | 1 | temporary table
After change | 1 | 0 | Base table
After change | 1 | 1 | not allowed/not used

an3l commented:
Closing this PR, since MDEV is pushed in 11.2
